### PR TITLE
Add Thread and ClassLoader Metrics for the self-observability and otel-oc-rules

### PR DESCRIPTION
**Pull request by Switch-vov**
### <Feature description>
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [x] If it's UI related, attach the screenshots below.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

**Review by wu-sheng — COMMENTED** [review]
You also should update OC receiver MAL settings, and grafana settings

**Comment by Switch-vov** [comment]
image

image

image

image

**Comment by Switch-vov** [comment]
> You also should update OC receiver MAL settings, and grafana settings

ok.

**Comment by wu-sheng** [comment]
FYI @wankai123 @Ax1an Please review this.

**Comment by Switch-vov** [comment]
@wu-sheng Only update OC receiver MAL settings. 
Because I don't known how generate the grafana settings `docs/en/setup/backend/grafana-cluster.json` and `docs/en/setup/backend/grafana-instance.json`.

Please teach me how to do it.😄

**Comment by wu-sheng** [comment]
Grafana UI provides that. @Ax1an Please confirm whether we have these already.

**Comment by Ax1an** [comment]
@wu-sheng Yes, our Grafana UI already has these.

**Comment by Ax1an** [comment]
@Switch-vov Please recheck the tests, it fails somehow.

**Comment by wu-sheng** [comment]
> @wu-sheng Yes, our Grafana UI already has these.

@Switch-vov You don't need to take care of Grafana. It is there. Please follow other comments.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml [hunk 1]
```diff
@@ -184,6 +184,42 @@ templates:
...
+                  "width": 3,
+                  "title": "Thread Count",
+                  "height": "200",
+                  "entityType": "ServiceInstance",
+                  "independentSelector": false,
+                  "metricType": "REGULAR_VALUE",
+                  "metricName": "meter_oap_instance_jvm_thread_live_count,meter_oap_instance_jvm_thread_daemon_count,meter_oap_instance_jvm_thread_peak_count",
+                  "queryMetricType": "readMetricsValues",
+                  "chartType": "ChartLine",
+                  "unit": "Per Minute"
```
Comment (wankai123) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml:198]: what does the unit `Per Minute`  mean?  If you want to get the time duration metric you should use like `rate('PT1M')`  or `increase('PT1M')` in the expressions.

**Comment by Switch-vov** [comment]
> @Switch-vov Please recheck the tests, it fails somehow.

ok.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml [hunk 2]
```diff
@@ -184,6 +184,42 @@ templates:
...
+                  "width": 3,
+                  "title": "Thread State Count",
+                  "height": "200",
+                  "entityType": "ServiceInstance",
+                  "independentSelector": false,
+                  "metricType": "REGULAR_VALUE",
+                  "metricName": "meter_oap_instance_jvm_thread_runnable_state_thread_count,meter_oap_instance_jvm_thread_blocked_state_thread_count,meter_oap_instance_jvm_thread_waiting_state_thread_count,meter_oap_instance_jvm_thread_timed_waiting_state_thread_count",
+                  "queryMetricType": "readMetricsValues",
+                  "chartType": "ChartBar",
+                  "unit": "Per Minute"
```
Comment (wankai123) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml:210]: also here.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml [hunk 3]
```diff
@@ -184,6 +184,42 @@ templates:
...
+                  "width": 3,
+                  "title": "Class Count",
+                  "height": "200",
+                  "entityType": "ServiceInstance",
+                  "independentSelector": false,
+                  "metricType": "REGULAR_VALUE",
+                  "metricName": "meter_oap_instance_jvm_class_loaded_class_count,meter_oap_instance_jvm_class_total_unloaded_class_count,meter_oap_instance_jvm_class_total_loaded_class_count",
+                  "queryMetricType": "readMetricsValues",
+                  "chartType": "ChartArea",
+                  "unit": "Per Minute"
```
Comment (wankai123) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml:222]: and here

**Review by wankai123 — CHANGES_REQUESTED** [review]

**Review by wankai123 — DISMISSED** [review]
LGTM

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml (hunk 1 above)
Comment (Switch-vov) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml]: done

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml (hunk 2 above)
Comment (Switch-vov) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml]: done

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml (hunk 3 above)
Comment (Switch-vov) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml]: done

**Comment by Switch-vov** [comment]
> @Switch-vov Please recheck the tests, it fails somehow.

In My IDEA, I try to run the `org.apache.skywalking.e2e.storage.StorageE2E#so11y`.

environment variables: `SW_STORAGE=postgresql; SW_AGENT_JDK_VERSION=8`

The result is:
image

I just add some values in `org.apache.skywalking.e2e.metrics.MetricsQuery#ALL_SO11Y_LINER_METRICS`.
It only used by `org.apache.skywalking.e2e.SO11yE2E#so11y` in Storage E2E.

@Ax1an So, Can you help me resolve it?

**Comment by Ax1an** [comment]
image

image

image

FYI @Switch-vov 

@kezhenxu94 I'm not very familiar with our testing framework. Can you help him see it?

File: test/e2e/e2e-data/src/main/java/org/apache/skywalking/e2e/metrics/MetricsQuery.java [hunk 4]
```diff
@@ -137,13 +137,21 @@
...
+    public static String METER_INSTANCE_JVM_CLASS_TOTAL_LOADED_CLASS_COUNT = "meter_oap_instance_jvm_class_total_loaded_class_count";
 
     public static String[] ALL_SO11Y_LINER_METRICS = {
         METER_INSTANCE_CPU_PERCENTAGE,
         METER_INSTANCE_JVM_MEMORY_BYTES_USED,
         METER_INSTANCE_METRICS_FIRST_AGGREGATION,
         METER_INSTANCE_PERSISTENCE_PREPARE_COUNT,
-        METER_INSTANCE_PERSISTENCE_EXECUTE_COUNT 
+        METER_INSTANCE_PERSISTENCE_EXECUTE_COUNT,
+        METER_INSTANCE_JVM_THREAD_LIVE_COUNT,
```
Comment (wu-sheng) [inline test/e2e/e2e-data/src/main/java/org/apache/skywalking/e2e/metrics/MetricsQuery.java:151]: Does the live thread metric really have values? I can't see that from your screenshots.

Comment (wu-sheng) [reply test/e2e/e2e-data/src/main/java/org/apache/skywalking/e2e/metrics/MetricsQuery.java:151]: Or other metrics are always having no-zero values? E2e is checking the values of metrics.

**Comment by wankai123** [comment]
image
This e2e seems ok

File: test/e2e/e2e-data/src/main/java/org/apache/skywalking/e2e/metrics/MetricsQuery.java (hunk 4 above)
Comment (Switch-vov) [reply test/e2e/e2e-data/src/main/java/org/apache/skywalking/e2e/metrics/MetricsQuery.java:151]: https://github.com/apache/skywalking/pull/7367#issuecomment-885995038

img

``` yaml
  - name: instance_jvm_thread_live_count
    exp: jvm_threads_current.sum(['service', 'instance'])
```

```
"metricName": "meter_oap_instance_jvm_thread_live_count,meter_oap_instance_jvm_thread_daemon_count,meter_oap_instance_jvm_thread_peak_count"
```

**Comment by wankai123** [comment]
image
SelfObservability e2e seems ok
image
Storage e2e has container issue ?

**Comment by wu-sheng** [comment]
I have rerun tasks, but still fail.
I don't know whether too many self observability metrics cost all left resources.

**Comment by wu-sheng** [comment]
@Switch-vov I noticed there is a new PostgreSQL 13 tag updated in dockerhub,  I rolled back to 13.2, https://hub.docker.com/layers/postgres/library/postgres/13.2/images/sha256-933b3327fd8545b1ea60dcf04c38fe6c4a974d48d73aa3771182f21d5e9ad68e?context=explore, which is released in 2 months ago.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml [hunk 5]
```diff
@@ -184,6 +184,39 @@ templates:
...
+                  "chartType": "ChartLine"
+                },
+                {
+                  "width": 3,
+                  "title": "Thread State Count",
+                  "height": "200",
+                  "entityType": "ServiceInstance",
+                  "independentSelector": false,
+                  "metricType": "REGULAR_VALUE",
+                  "metricName": "meter_oap_instance_jvm_thread_runnable_state_thread_count,meter_oap_instance_jvm_thread_blocked_state_thread_count,meter_oap_instance_jvm_thread_waiting_state_thread_count,meter_oap_instance_jvm_thread_timed_waiting_state_thread_count",
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/self-observability.yml:206]: Could you run test on real PostgreSQL? I am feeling whether this name, `meter_oap_instance_jvm_thread_runnable_state_thread_count`, is too long for PostgreSQL.

File: oap-server/server-bootstrap/src/main/resources/fetcher-prom-rules/self.yaml [hunk 6]
```diff
@@ -79,3 +79,23 @@ metricsRules:
...
+  - name: instance_jvm_thread_waiting_state_thread_count
+    exp: jvm_threads_state.tagMatch('state', 'WAITING').sum(['service', 'instance'])
+  - name: instance_jvm_thread_timed_waiting_state_thread_count
+    exp: jvm_threads_state.tagMatch('state', 'TIMED_WAITING').sum(['service', 'instance'])
+  - name: instance_jvm_class_loaded_class_count
+    exp: jvm_classes_loaded.sum(['service', 'instance'])
+  - name: instance_jvm_class_total_unloaded_class_count
+    exp: jvm_classes_unloaded_total.sum(['service', 'instance'])
+  - name: instance_jvm_class_total_loaded_class_count
+    exp: jvm_classes_loaded_total.sum(['service', 'instance'])
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/fetcher-prom-rules/self.yaml:101]: Consider removing `instance_` as a prefix, as JVM has indicated it is instance level.
For example, change `instance_jvm_thread_waiting_state_thread_count` to `jvm_thread_waiting_count`. Remove unnecessary words. The different storages have different rules.

**Comment by wu-sheng** [comment]
> > @Switch-vov Please recheck the tests, it fails somehow.
> 
> In My IDEA, I try to run the `org.apache.skywalking.e2e.storage.StorageE2E#so11y`.
> 
> environment variables: `SW_STORAGE=postgresql; SW_AGENT_JDK_VERSION=8`
> 
> The result is:
> image
> 
> I just add some values in `org.apache.skywalking.e2e.metrics.MetricsQuery#ALL_SO11Y_LINER_METRICS`.
> It only used by `org.apache.skywalking.e2e.SO11yE2E#so11y` in Storage E2E.
> 
> @Ax1an So, Can you help me resolve it?

The error is not on so11y, it is in the storage and ttl cases. Also relate to postgreSQL storage mostly.

**Comment by wu-sheng** [comment]
Please check whether there is namespace setting, which could make the metric table name longer.

**Comment by Switch-vov** [comment]
> Please check whether there is namespace setting, which could make the metric table name longer.

ok. I will rename and test it later.

**Comment by Switch-vov** [comment]
> > Please check whether there is namespace setting, which could make the metric table name longer.
> 
> ok. I will rename and test it later.

After simplify metrics name.

Run the `org.apache.skywalking.e2e.storage.StorageE2E`.

environment variables: `SW_STORAGE=postgresql; SW_AGENT_JDK_VERSION=8`

The result is:
image

I think some metric too long for PostgreSQL. Because the default max length of PostgreSQL is 64.

```c
src/include/pg_config_manual.h
#define NAMEDATALEN 64
```

**Comment by wu-sheng** [comment]
I can see the CI process has confirmed this guess.

**Comment by wu-sheng** [comment]
@wankai123 Could you run this locally to verify UI is as expected?

**Comment by Switch-vov** [comment]
thread_count

thread_state_count

class_count

**Comment by wankai123** [comment]
> @wankai123 Could you run this locally to verify UI is as expected?

It looks ok.
image

**Review by wu-sheng — APPROVED** [review]
Forward @wankai123 's approval, as his committer account has not been created by INFRA.

**Comment by Switch-vov** [comment]
fix value of `METER_INSTANCE_JVM_CLASS_TOTAL_LOADED_CLASS_COUNT`

**Review by wu-sheng — APPROVED** [review]

**Comment by Switch-vov** [comment]
Change some metrics name in `ALL_SO11Y_LINER_METRICS`

**Comment by wu-sheng** [comment]
Wait for CI confirm.

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**